### PR TITLE
Add support for EnergyModelsInvestments as extension

### PR DESCRIPTION
**Pull request by hellemo**
This is related to ongoing work to make `EnergyModelsInvestments` (EMI) independent of `EnergyModelsBase` (EMB).
This is a first working draft to discuss. This PR makes the following changes to EMB:

1. Relax type constraints on modeltype to not force extensions to depend on EMB
2. Relax type constraints on data to not force extensions to depend on EMB
3. Move EMB specific code from EMI to a new extension of EMB (`EMIExt`)
4. Format code using `JuliaFormatter`

Things to do:

- [x]  Reintroduce EnergyModel dependency for all functions. 
- [x]  Reintroduce the dependency of the data as data::Vector{Data} for the nodes. 
- [x]  Check all links in the documentation to point towards the proper directory in EMI. 
- [x]  Keep more original/manual formatting of constraints (and variables)

### CODE REVIEW

**Comment by JulStraus** [comment]
I moved the EMB specific types to the EMB extensions. This includes as well all legacy constructors, except for `TransInvData`.

This has two major implications. Some functions we used previously in the `GeoExt` are now required to be defined in the new extension in `EnergyModelsGeography`, namely `EMI.investment_data(tm::EMG.TransmissionMode)` , `has_investment(tm::EMG.TransmissionMode)`, and `EMI.start_cap(tm::EMG.TransmissionMode, t_inv, inv_data::NoStartInvData, cap)`.

The main reason to redefine them is to allow reference methods in `EnergyModelsInvestments` that provide an error if the function is not specified directly by the user.

Thinks to be fixed in this PR:

1. Add again the `EnergyModel` dependency for all functions. There is no reason to not have it. Any potential problem should be solved the same way as it is solved within the investment extension.
2. Add again the dependency of the data as `data::Vector{Data}` for the nodes. Again, we do not want the user to provide anything that is not a proper `EMB.Data` as that may result in unforeseen problems.
3. Check all links in the documentation to point towards the proper directory in `EMI`. I used a lot of "@ref" for a simplified cross referencing in the data types. I would say this is something we do entirely at the end.

I also want to reverse in the end some formatting from JuliaFormatter. In my opinion, constraint macro formatting is awful with JuliaFormatter as it does not capture the intrinsics of the mathematical equations.

**Comment by hellemo** [comment]
@JulStraus I added some options to JuliaFormatter to keep more of the manual formatting. Check `model.jl` and `constraint_functions.jl` for some examples where constraints are formatted manually before formatting with JuliaFormatter. I think this will satisfy most if not all concerns when it comes to formatting of constraints.

**Comment by JulStraus** [comment]
> @JulStraus I added some options to JuliaFormatter to keep more of the manual formatting. Check `model.jl` and `constraint_functions.jl` for some examples where constraints are formatted manually before formatting with JuliaFormatter. I think this will satisfy most if not all concerns when it comes to formatting of constraints.

Generally speaking, it looks better now. An alternative could also to use [`ignore`](https://domluna.github.io/JuliaFormatter.jl/stable/config/#ignore) for `constraint_functions.jl` and `data_functions.jl`, although I am not fully certain that it would improve it.

File: docs/src/how-to/create-new-node.md [hunk 1]
```diff
@@ -23,32 +28,32 @@ This is however only advised if you do not need to access the value of the varia
...
 2. Emissions can be included in any way.
    It is however beneficial to reutilize the [`EmissionsData`](@ref) type to improve usability with other packages.
    This requries again the inclusion of the field `data` in `NewNodeType`.
-   It is possible to also create new subtypes for `EmissionsData` as well as dispatch on the function [`constraints_data(m, n::Node, 𝒯, 𝒫, modeltype, data::Data)`](@ref data_functions).
+   It is possible to also create new subtypes for `EmissionsData` as well as dispatch on the function [`constraints_data(m, n::Node, 𝒯, 𝒫, modeltype, data::Data)`](@ref man-data_fun).
 3. It is in a first stage not important to include functions for handling all possible `TimeStructure`s, that is, *e.g.*, `RepresentativePeriods`.
    Instead, and error can be provided if an unsupported `TimeStructure` is chosen.
-4. The existing reference nodes and their respectvve *[constraint functions](@ref constraint_functions)* can serve as idea generators.
+4. The existing reference nodes and their respectvve *[constraint functions](@ref man-con)* can serve as idea generators.
```
Comment (hellemo) [inline docs/src/how-to/create-new-node.md:45]:
```suggestion
4. The existing reference nodes and their respective *[constraint functions](@ref man-con)* can serve as idea generators.
```

**Pull request merged**